### ISSUE TITLE
Path-related methods on `file` type

### ISSUE THREAD

**Issue by e-kayrakli** [chapel-lang/chapel#16748]
We have the following methods on `file` defined in the `Path` module:

- `file.absPath`
- `file.getParentName`
- `file.realPath`
- `file.relPath`

But the following functions in the `Path` module doesn't define symmetric
methods on `file`:

- `splitPath`
- `normPath`
- `basename`

This raises the following questions:
- Why do we have methods on file for some Path functions but not for some?
- Should these be in the Path module?

Few strategies that seem to have supporters:

1. Move `file.absPath` and `file.realPath` to IO, make them independent of
   `Path` by storing relevant information in the `file` type itself.

2. Remove them altogether. This would make users do `use Path;
   absPath(file.path);`, for example.

3. Keep them in the Path module, but re-arrange what is available on `file`: it
   doesn't make much sense to have `getParentName` (or `dirname`) on `file`, but
   not `basename`, for example.  That doesn't mean we should add
   `file.splitPath`, however.

Notes:
- Motivation for having methods on `file` can be relying on different
  implementations and information stored in `file`. However, the current
  implementation of `file` doesn't lend itself to that readily. So, there may be
  more implementation on that side for that to happen.

  One similar motivation is to avoid using cwd or other means that can be
  race-y.

- @ben-albrecht noted that Python's file type doesn't have path-specific methods

Path module reference: https://chapel-lang.org/docs/master/modules/standard/Path.html

**Comment by vasslitvinov** [comment]
I vote for (1). I think Michael implied that no additional information would need to be stored in `file` for that.

**Comment by ben-albrecht** [comment]
> @ben-albrecht noted that Python's file type doesn't have path-specific methods

beyond `file.name` which is equivalent to out `file.path` (can be abs path or relative path depends on how `file` was created)

**Comment by lydia-duncan** [comment]
My vote is for (3).  I already said this in the meeting but I want the implementations:
- unified
- sorted near each other
and I don't see the harm of continuing to maintain a helper method that a user would probably write themselves otherwise.

**Comment by bradcray** [comment]
> Why do we have methods on file for some Path functions but not for some?

My answer to this question would be that the ideal would be to have the `file.` type support methods that relate to getting the path corresponding to a file (i.e., things that might be considered a property of the file), but not for doing string processing on that path.  This is why it seems reasonable to me that some, but not all path modules be supported on files.  

I bought Michael's arguments that those methods whose results depend on the value of cwd don't feel as much like they're part of a file's definition (since they're a function of both file path and cwd) and therefore should potentially be removed.

That said, I couldn't tell you the difference between `file.path` and `file.*path()`.  Based on today's discussion, I'm under the impression that a file should store an absolute, normalized path rather than relative or non-normalized one within its state (whether as one string or two), but that this isn't necessarily the case today (?).

All that said, given that the Path module follows Python's design, is the implication that Python supports these four methods on files as well?  Oh wait, I just saw this:

> @ben-albrecht noted that Python's file type doesn't have path-specific methods

So Lydia, do you remember what it was that caused us to add these rather than follow Python's lead?

**Comment by lydia-duncan** [comment]
> So Lydia, do you remember what it was that caused us to add these rather than follow Python's lead?

This is from the original proposal document:

> Through this proposal, we seek to expand both the file record capabilities and
the manipulation of files via string specifiers.  Most of the current
functionality will remain unchanged, with the exception of the
file.path/getPath/tryGetPath methods which will be both expanded and subsumed
by the functions absPath, realPath, and relPath and their respective file
methods.

Which to me sounds like generally adding methods to file was part of the stated goal of the FileSystem and Path libraries from our perspective.  I definitely don't remember it being controversial at the time (though that's not to say that we can't change our mind).

**Comment by e-kayrakli** [comment]
I am for (1).

I started with (2) in mind, and as the discussion goes, it feels more and more like (1) and (2) are more or less the same thing, where (1) is just a better characterization of what needs to be done.

> My answer to this question would be that the ideal would be to have the file. type support methods that relate to getting the path corresponding to a file (i.e., things that might be considered a property of the file), but not for doing string processing on that path. This is why it seems reasonable to me that some, but not all path modules be supported on files.

Michael's comments and this resonate with me. And it follows from this that if such methods are only about getting a property of a file, they should be defined where the file type is defined (the IO module) instead of the Path module.

**Comment by ben-albrecht** [comment]
I was originally for (2), but if the file methods are reduced to simply `file.absPath()` and `file.path` (redundant with `file.realPath`), I'd be fine (1).

It might make sense to convert `file.absPath()` to a field similar to `file.path`, but that question doesn't have to be tackled here.

**Comment by mppf** [comment]
Others have already brought up my argument for 1:
 * Unlike the general `string.absPath` `file.absPath()` need not use `cwd` (generally speaking)
 * `file.absPath` and `file.realPath` are arguably variant ways of getting the path from the file rather than string processing applied after the fact.

I was hoping that `file.absPath` and `file.realPath` could be implemented without any of the `Path` module functionality but I think they do normalize the path (which can/should be shared with other code in Path).

> That said, I couldn't tell you the difference between file.path and file.*path(). Based on today's discussion, I'm under the impression that a file should store an absolute, normalized path rather than relative or non-normalized one within its state (whether as one string or two), but that this isn't necessarily the case today (?).

We are getting the file path from the OS file descriptor. On linux this is done by reading `/proc/self/fd/#` which apparently stores absolute paths. On Mac OS X we use fcntl with `F_GETPATH` which is also apparently absolute.

So, these start out absolute but we make them relative again (if shorter) with `qio_shortest_path` in the implementation of `file.path`.

Anyway, the comment above makes me think it makes more sense to put `file.path`, `file.absPath`, and `file.realPath` next to each other in the file documentation, so that it's clear which variant of the path one is getting / which variants are available.

**Comment by e-kayrakli** [comment]
@mppf said:

> I was hoping that file.absPath and file.realPath could be implemented without any of the Path module functionality but I think they do normalize the path (which can/should be shared with other code in Path).



> We are getting the file path from the OS file descriptor. On linux this is done by reading /proc/self/fd/# which apparently stores absolute paths. On Mac OS X we use fcntl with F_GETPATH which is also apparently absolute.

This is right, but, do we want to be doing those things?

Consider the structure below:
```
 $ ls -gG filePathTest/

baseFile
myLink -> ./../filePathTest/baseFile

```

where `myLink` is a symbolic link created with a non-normal path. If we have a
file object opened on that link, we never get the link when we do `file.path`,
as we ask the OS for the path to a file object (and that would end up calling
`readlink` under the covers), which gives an absolute, normalized and real path.

<details>

```chapel
use IO;
use Path;


{
  writeln("Normal relative path was given");
  var f = open("filePathTest/myLink", iomode.r);
  writeln("f.path: ", f.path);
  writeln("f.absPath: ", f.absPath());
  writeln("f.realPath: ", f.realPath());
  writeln("f.relPath: ", f.relPath());
  writeln();
  f.close();
}


{
  writeln("Non-Normal relative path was given");
  var f = open("./filePathTest/myLink", iomode.r);
  writeln("f.path: ", f.path);
  writeln("f.absPath: ", f.absPath());
  writeln("f.realPath: ", f.realPath());
  writeln("f.relPath: ", f.relPath());
  writeln();
  f.close();
}

{
  writeln("Absolute path was given");
  var f = open("/absolute/path/filePathTest/myLink", iomode.r);
  writeln("f.path: ", f.path);
  writeln("f.absPath: ", f.absPath());
  writeln("f.realPath: ", f.realPath());
  writeln("f.relPath: ", f.relPath());
  writeln();
  f.close();
}

```

Gives:

```
Normal relative path was given
f.path: filePathTest/baseFile
f.absPath: /absolute/path/filePathTest/baseFile
f.realPath: /absolute/path/filePathTest/baseFile
f.relPath: filePathTest/baseFile

Non-Normal relative path was given
f.path: filePathTest/baseFile
f.absPath: /absolute/path/filePathTest/baseFile
f.realPath: /absolute/path/filePathTest/baseFile
f.relPath: filePathTest/baseFile

Absolute path was given
f.path: filePathTest/baseFile
f.absPath: /absolute/path/filePathTest/baseFile
f.realPath: /absolute/path/filePathTest/baseFile
f.relPath: filePathTest/baseFile
```

This behavior is different than python (which only has `file.name` on the file
object:

```python
with open("filePathTest/myLink", "r") as f:
  print(f.name)
  print()

with open("./filePathTest/myLink", "r") as f:
  print(f.name)
  print()

with open("/absolute/path/filePathTest/myLink") as f:
  print(f.name)
  print()
```

Gives:

```
filePathTest/myLink

./filePathTest/myLink

/absolute/path/filePathTest/myLink
```

</details>

So, there's no way to get the file path as in "the string used to create the
file object". And this is bad because I think it'd be important to get that
information, and with the status quo, some of these methods on file type are
redundant.

---

On this matter @ben-albrecht said:

> There is a file.name field which is equivalent to out file.path (can be abs
> path or relative path depends on how file was created)

So, this is not entirely correct. However, I think this should be our design.
i.e. `file.path` should return the string we used to open the file.

---

So, maybe the solution is to store some information in the `file` type fields, 
(@vasslitvinov -- [do you want to avoid adding new fields to `file`?](https://github.com/chapel-lang/chapel/issues/16748#issuecomment-730651661))
and limit the methods to be getters for those fields. I propose we take the
following steps:

1. Add "private" `path` and `dirname` fields to `file`. And add getters for
   them.

   These getters will automatically replace `file.path` and `file.getParentName`
   in the Path module.

2. Move `realPath` and `absPath` to the IO module. This'll make them available
   when someone does `use IO`, and `use Path` will no longer be required for
   them.

   I can think of two objections to this:

  - Two separate implementations for pretty much the same thing.
    (`Path.realPath` vs `file.realPath`)

    2a. I think we can have an `internal/ChapelFS.chpl` module that has thin
    wrappers for our filesystem support at runtime. Then, this module can be
    used from IO, Path and FileSystem modules. This would make the
    implementation of `realPath` and `absPath` on both modules to be 1-liners,
    more or less.

  - Having a `realPath()` and `file.realPath()` can be confusing. Especially
    so, if the latter is available without `use Path;`

    2b. We can make `file.realPath` and `file.absPath` paren-less methods, so,
    interface-wise they look just like properties on `file` type rather than
    methods. This'll also make things consistent within the `file` interface:
    `file.path`, `file.dirname`, `file.realPath`, `file.absPath`

3. Drop `file.relPath` altogether

  I personally see this one as less related to the `file` type. It returns the
  relative path to the file from a given path, which defaults to `.`

**Comment by mppf** [comment]
I'd be fine with that general direction. In that event, I have a slight preference to leave out `file.realPath` and `file.absPath`, since I would think that `realPath(myFile.path)` is not much harder to write that `file.realPath`. However it is not a big deal either way and (as the discussion has shown) my opinion for which of `file.path`, `file.realPath`, `file.absPath` we should have depends on the implementation. Providing all 3 would be a way to protect us from that dependency. (However what you are describing with `file.path` and symbolic links pretty much locks us in to storing the path in the `file` somewhere. My view is that if we do store it in `file` then we should store it within the C structure backing it. That way `file` continues to be really just a pointer to an object with details. But, this is not strictly required.)

**Comment by e-kayrakli** [comment]
> I'd be fine with that general direction. In that event, I have a slight preference to leave out file.realPath and file.absPath, since I would think that realPath(myFile.path) is not much harder to write that file.realPath.

I think `realPath(myFile.path)` would only work if the file was created with an absolute path, or if we haven't changed `cwd` since creating the file.

A more robust way of getting realPath might be to get the absolute realPath by `realPath(joinPath(myFile.dirname, myFile.path))`. (You can also call `relPath` on the returned value to get the relative realPath, but you get the idea)

**Comment by vasslitvinov** [comment]
If we expect that a file's path (of whatever variety) will continue to be available in the C struct, I suggest we expose it as a method on `file`, i.e. without the Path module. I suggest labeling it as implementation-dependent whether it is absolute, relative, or whatever. The user can polish this path using appropriate `Path` functions.

**Comment by e-kayrakli** [comment]
After today's meeting, we are thinking of taking the following steps:

1. deprecate `file.realPath`, `file.relPath` and `file.absPath`, in favor of standalone overloads with file arguments: `realPath(file)`, `relPath(file)`, `absPath(file)`
2. deprecate `file.getParentName` for good

Several things led to this direction:

- `realPath` does operations on the file descriptor that cannot be easily done by the user.
- The other two are similar enough to `realPath` that we want to keep them as siblings (as standalone functions)
- We want these implementations to be close (as in location in source) to their cousins that operate on `string` arguments
- There are some concerns about having them implemented as tertiary methods:
  - Documenting them in a single place was one, with the caveat that we have been thinking about a "type view" for chpldoc, which would help the case
  - General/philosophical concerns about having tertiary methods in the standard modules
- As for `file.getParentName`: the obvious alternative would be `dirname(file)`. However, this isn't very clear as to what type of path (absolute, relative...) it's operating on. Moreover, there's not much motivation for keeping it. It seems that the name is going to deprecate anyways, and we can always consider adding `dirname(file)` later on.

Potential cons:

- `file.path` and `realPath(file)` doesn't feel very symmetric. However, `realPath` is going to be in the `Path` module, and within that interface we'll have symmetry between `realPath(file)` and `realPath(string)`
- The standalone `realPath(file)` will have the fiddle with the internals of the file type, and that's not ideal. We can think of adding a method on `file` that can give the file descriptor, and a function in `FileSystem` that can get real path from a descriptor. If we have those user-facing features, `Path.realPath(file)` could use those.

Other matters:

- We can think of adding file-argument overloads for other `Path` module for symmetry. We don't have a lot of motivation to do so, and if we do, it is not a breaking change.

**Comment by bradcray** [comment]
I'll be curious for @gbtitus's opinion on the previous post when he's back from vacation, since I believe he had some comments on the placement of these methods in the original review.

Regarding:

> General/philosophical concerns about having tertiary methods in the standard modules

I don't know that I inherently object to using tertiary methods in standard modules, but the argument that appealed most to me here was Michael's:  "I get a file object from somewhere, and it feels like I should be able to do all of the standard library methods on files on it, but I can't call `.absPath()` on it (say) without doing `use Path;` or `import Path.file;` which seems frustratingly odd compared to all of the other methods on file."

**Comment by jhh67** [comment]
After giving this more thought I think we should simply provide the path that was used to open the file. Everything else has the potential to be stale as soon as it is returned, and I would lean toward not providing any information vs. information that might be incorrect. As an alternative, I would concede returning the absolute path used to open the file as long as the open function converts the provided path to an absolute path before opening the file (although this too introduces a race unless there is a way to convert to an absolute path and open a file atomically that I'm not aware of). Posix doesn't provide a way of returning the current path of an open file because of the difficulty in doing so and because there can be 0 or more answers. There are OS-specific ways of doing it (e.g. the /proc filesystem) so it's doable, but not portable. I have no opinion on the tertiary method issue -- Brad explained it but I don't fully grok it yet.

**Comment by bradcray** [comment]
It seems like there are lots of queries about the outside world that one might make (or want to make) in Chapel whose results could be invalid by the time they're returned (from "What's the size of this file?" to "What files are in this directory?" to "How many tasks are running?" to "What time is it?").  So, to me, the fact that something is changing or could change doesn't feel like sufficient grounds to not support a query.

While I understand that these things can change asynchronously, it also seems like it's not difficult to create well-behaved cases like "My program opens a file in a directory that only I have read/write access to, and I promise myself that my program and interactive shell won't interact with the file while it's doing so."  Should I be prevented from making such queries on my files in such cases?  (i.e., could the documentation just state the potential for the result to be outdated and leave it to users to decide whether they can trust such information or not).  Of course, I suppose it could be argued that if the system is that much under my control, maybe I don't even need the queries to begin with (?).

**Comment by mppf** [comment]
FWIW my view is that the path used to open the file can also become stale and getting the path from the OS is more robust and less likely to be stale, but it is true that neither is necessarily what the user wants in all cases. (Of course we could define the query as returning "the path used to open the file" but I'm not sure that in itself is more useful than "the current path to the file according to the OS" which is the current behavior).

**Comment by jhh67** [comment]
All good points. Clearly I need to give it more thought. Somehow I think of the file name(s) differently from the other attributes of a file in ways I can't quite explain. I'll note that the path used to open the file is immutable. It doesn't become stale because the path used to open the file can't change.

**Comment by vasslitvinov** [comment]
My view on tertiary methods is that they bring us back to visibility control based on modules, rather than classes.

In other words, if I need to import module `M` to be able to call a function `xyz()`, it makes sense that I need to import module `N` to be able to call a method `x.yz()`.

This makes tertiary methods not a consideration for me in this case.

**Comment by gbtitus** [comment]
All file characteristics are unstable in the sense that they can change over time, whether or not you have the file open. File paths still seem slightly different to me than other characteristics of files such as modes (protections) and sizes, however. The latter are all things having to do with the file itself; a file has precisely one mode and size even though those may vary with time. But a file doesn't necessarily have just one path. Note that for example the POSIX `realpath(3)` function is documented as producing "**_an_** absolute pathname that names the [...] file" (emphasis mine), not "**_the_**" absolute pathname. That's because even without any relative components or symbolic links, mount points and hard links can still create a situation in which you can reach a single file via more than one path. So while there is necessarily "the size" of a file, there is not necessarily "the path" to it.

I'm not sure how much this helps with deciding what to do here, though.  🤓

**Comment by e-kayrakli** [comment]
What I am understanding is that there's no great way of getting the path (if there's any, that is) of a file, and we are in general OK with the current behavior of pertinent file methods in the Path module.

I am also not seeing any objection to making those methods into functions, which https://github.com/chapel-lang/chapel/pull/18009 is going to do.

I plan to merge that tomorrow (Thursday), unless I am misinterpreting the discussion here.

I can open a spinoff issue for the final part of the discussion, though.

**Comment by e-kayrakli** [comment]
This issue is resolved via #18009
